src/app/components/Loop/index.tsx

```diff
@@ -0,0 +1,425 @@
+import { Form, Button, Icon, Radio, Tooltip, Input, Switch, message } from 'antd';
+import AmountField from 'components/AmountField';
+import InputLoopAddress from 'components/Loop/InputLoopAddress';
+import QuoteModal from './QuoteModal';
+import { Menu, Dropdown } from 'antd';
+import { ChannelWithNode } from 'modules/channels/types';
+
+interface StateProps {
+  channels: AppState['channels']['channels'];
+  url: AppState['loop']['url'];
+  lib: AppState['loop']['lib'];
+  loopTerms: AppState['loop']['loopTerms'];
+  loopQuote: AppState['loop']['loopQuote'];
+  loop: AppState['loop']['loop'];
+  error: AppState['loop']['error'];
+}
+
+interface DispatchProps {
+  getLoopOutTerms: typeof getLoopOutTerms;
+  getLoopInTerms: typeof getLoopInTerms;
+  setLoop: typeof setLoop;
+  setLoopIn: typeof setLoopIn;
+}
+
+interface State {
+  amount: string;
+  advanced: boolean;
+  isAnyValue: boolean;
+  destination: string;
+  swapRoutingFee: string;
+  swapFee: string;
+  minerFee: string;
+  prepayAmt: string;
+  channel: string;
+  conf: string;
+  htlc: boolean;
+  quoteModalIsOpen: boolean;
+  loopType: string;
+}
+
+const INITIAL_STATE = {
+  amount: '0',
+  advanced: false,
+  isAnyValue: false,
+  destination: '',
+  swapRoutingFee: '',
+  swapFee: '0',
+  minerFee: '0',
+  prepayAmt: '0',
+  channel: '',
+  conf: '2',
+  htlc: false,
+  quoteModalIsOpen: false,
+  loopType: 'Loop Out',
+};
+
+type Props = StateProps & DispatchProps;
+
+class Loop extends React.Component<Props> {
+  state: State = { ...INITIAL_STATE };
+
+  render() {
+    const { url, loopTerms, channels, error } = this.props;
+
+    if (channels === null || loopTerms === null) {
+      return null;
+    }
+
+    // Only return channels with enough loot for looping out
+    const openChannelsLoopOut = channels.filter(
+      o =>
+        o.status === 'OPEN' &&
+        parseInt(o.local_balance, 10) > parseInt(loopTerms.min_swap_amount, 10),
+    );
+    // Only return channels with enough loot for looping in
+    const openChannelsLoopIn = channels.filter(
+      o =>
+        o.status === 'OPEN' &&
+        parseInt(o.capacity, 10) - parseInt(o.local_balance, 10) >
+          parseInt(loopTerms.min_swap_amount, 10),
+    );
+
+    // Get channels to choose from
+    const loopOutItems = openChannelsLoopOut.map(c => (
+      <Menu.Item
+        key={c.channel_point}
+        onClick={() => this.handleSetChannelId(c.channel_point, openChannelsLoopOut)}
+      >
+        {`${c.node.alias} => ${c.local_balance} sats available`}
+      </Menu.Item>
+    ));
+    const loopInItems = openChannelsLoopIn.map(c => (
+      <Menu.Item
+        key={c.channel_point}
+        onClick={() => this.handleSetChannelId(c.channel_point, openChannelsLoopIn)}
+      >
+        {`${c.node.alias} => ${c.local_balance} sats available`}
+      </Menu.Item>
+    ));
+
+    const loopInMenu = <Menu>{loopInItems}</Menu>;
+    const loopOutMenu = <Menu>{loopOutItems}</Menu>;
```

Comment:
These behave really poorly as `<Menu>`s (especially if you have a lot of channels.) I think they should be `<Select>` lists instead.

Reply:
The Select is looking much better for sure. It was super hacky before. Only problem I was seeing now is that console complains that I should use Select.Option, but then typescript complains that Select.Options has no onClick prop. So I left as Menu.Item